File: 520.detect-capital/520.detect-capital.cpp
/*
 * @lc app=leetcode id=520 lang=cpp
 *
 * [520] Detect Capital
 *
 * https://leetcode.com/problems/detect-capital/description/
 *
 * algorithms
 * Easy (53.05%)
 * Likes:    477
 * Dislikes: 251
 * Total Accepted:    112.7K
 * Total Submissions: 211.1K
 * Testcase Example:  '"USA"'
 *
 * Given a word, you need to judge whether the usage of capitals in it is right
 * or not.
 * 
 * We define the usage of capitals in a word to be right when one of the
 * following cases holds:
 * 
 * 
 * All letters in this word are capitals, like "USA".
 * All letters in this word are not capitals, like "leetcode".
 * Only the first letter in this word is capital, like "Google".
 * 
 * Otherwise, we define that this word doesn't use capitals in a right way.
 * 
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: "USA"
 * Output: True
 * 
 * 
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: "FlaG"
 * Output: False
 * 
 * 
 * 
 * 
 * Note: The input will be a non-empty word consisting of uppercase and
 * lowercase latin letters.
 * 
 */

// @lc code=start
class Solution
{
public:
    bool detectCapitalUse(string word)
    {
        if (word.size() <= 1)
            return true;

        bool up = false;
        if (isUp(word[0]) && isUp(word[1]))
            up = true;

        for (int i = 1; i < word.size(); i++)
        {
            if ((up && !isUp(word[i])) || (!up && isUp(word[i])))
                return false;
        }

        return true;
    }

    bool isUp(char c)
    {
        return 'A' <= c && c <= 'Z';
    }
};
// @lc code=end

File: 520.detect-capital/520.detect-capital.java
/*
 * @lc app=leetcode id=520 lang=java
 *
 * [520] Detect Capital
 *
 * https://leetcode.com/problems/detect-capital/description/
 *
 * algorithms
 * Easy (54.20%)
 * Likes:    928
 * Dislikes: 309
 * Total Accepted:    203.7K
 * Total Submissions: 375.9K
 * Testcase Example:  '"USA"'
 *
 * We define the usage of capitals in a word to be right when one of the
 * following cases holds:
 * 
 * 
 * All letters in this word are capitals, like "USA".
 * All letters in this word are not capitals, like "leetcode".
 * Only the first letter in this word is capital, like "Google".
 * 
 * 
 * Given a string word, return true if the usage of capitals in it is right.
 * 
 * 
 * Example 1:
 * Input: word = "USA"
 * Output: true
 * Example 2:
 * Input: word = "FlaG"
 * Output: false
 * 
 * 
 * Constraints:
 * 
 * 
 * 1 <= word.length <= 100
 * word consists of lowercase and uppercase English letters.
 * 
 * 
 */

// @lc code=start
class Solution {
    public boolean detectCapitalUse(String word) {
        if (word == null || word.isEmpty()) {
            return true;
        }

        if (word.charAt(0) >= 'A' && word.charAt(0) <= 'Z') {
            return allUpper(word.substring(1)) || allLower(word.substring(1));
        }

        return allLower(word.substring(1));
    }

    private boolean allUpper(String s) {
        for (char c : s.toCharArray()) {
            if (c < 'A' || c > 'Z') {
                return false;
            }
        }

        return true;
    }

    private boolean allLower(String s) {
        for (char c : s.toCharArray()) {
            if (c < 'a' || c > 'z') {
                return false;
            }
        }

        return true;
    }
}
// @lc code=end

